reference/best-practices/java-app.md

```diff
@@ -0,0 +1,362 @@
+- Data access framework: Applications usually use a data access framework such as [MyBatis](https://mybatis.org/mybatis-3/index.html) and [Hibernate](https://hibernate.org/) to further simplify and manage the database access operations.
+- Application implementation: The application logic controls when to send what commands to the database. Some applications use [Spring Transaction](https://docs.spring.io/spring/docs/4.2.x/spring-framework-reference/html/transaction.html) aspects to manage transactions' begin and commit logics.
+
+![Java application components](/media/java-practice-1.png)
+
+As shown above, a Java application may do the following things:
+
+- Implement the MySQL protocol via the JDBC API to interact with TiDB
+- Get a persistent connection from the connection pool
+- Use a data access framework such as MyBatis to generate and execute SQL statements
+- Use Spring Transaction to automatically start or stop a transaction
+
+The rest of this document describes the issues and their solutions when you develop a Java application using the above components.
+
+## JDBC
+
+Java applications can be encapsulated with various frameworks. In most of the frameworks, JDBC API is called on the lowest level to interact with the database server. For JDBC, it is recommended that you focus on the following things:
+
+- JDBC API usage choice
+- API Implementer's parameter configuration
+
+### JDBC API
+
+For JDBC API usage, see [JDBC official tutorial](https://docs.oracle.com/javase/tutorial/jdbc/). This section covers the usage of several important APIs.
+
+#### Use Prepare API
+
+For OLTP (Online Transactional Processing) scenarios, the SQL statements sent by the program to the database are several types that can be exhausted after removing parameter changes. Therefore, it is recommended to use [Prepared Statements](https://docs.oracle.com/javase/tutorial/jdbc/basics/prepared.html) instead of regular [execution from a text file](https://docs.oracle.com/javase/tutorial/jdbc/basics/processingsqlstatements.html#executing_queries) and reuse Prepared Statements to execute directly. This avoids the overhead of repeatedly parsing and generating SQL execution plans in TiDB.
+
+At present, most upper-level frameworks call the Prepare API for SQL execution. If you use the JDBC API directly for development, pay attention to choosing the Prepare API.
+
+In addition, with the default implementation of MySQL Connector/J, only client-side statements are preprocessed, and the statements are sent to the server in a text file after `?` is replaced on the client. Therefore, in addition to using the Prepare API, you also need to configure `useServerPrepStmts = true` in JDBC connection parameters before you perform statement preprocessing on the TiDB server (the parameter configuration section is described in detail below).
```

Comment:
What about adding an anchor to locate the exact section?

Reply:
Okay.